// ngx-monaco-editor-v2/fesm2022/ngx-monaco-editor-v2.mjs
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, output, viewChild, effect, ChangeDetectionStrategy, Component, NgZone, forwardRef, NgModule, makeEnvironmentProviders } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { fromEvent } from 'rxjs';
import { CommonModule } from '@angular/common';

const NGX_MONACO_EDITOR_CONFIG = new InjectionToken('NGX_MONACO_EDITOR_CONFIG');

let loadedMonaco = false;
let loadPromise;
class BaseEditor {
    constructor() {
        this.config = inject(NGX_MONACO_EDITOR_CONFIG);
        /** Run monaco creation inside Angular's zone (defaults to outside for performance). */
        this.insideNg = input(false, /* @ts-ignore */
        ...(ngDevMode ? [{ debugName: "insideNg" }] : /* istanbul ignore next */ []));
        /** Emits the underlying monaco editor instance once it has been created. */
        this.onInit = output();
        this.editorContainer = viewChild.required('editorContainer');
        /** Models created by this component, tracked so they can be disposed (monaco keeps them globally otherwise). */
        this._models = [];
        /** Editor event listeners, disposed alongside the editor. */
        this._listeners = [];
        // Re-create the editor when `insideNg` changes after it already exists.
        effect(() => {
            this.insideNg();
            if (this._editor) {
                this.reinit();
            }
        });
    }
    ngAfterViewInit() {
        if (loadedMonaco) {
            // Wait until monaco editor is available
            loadPromise.then(() => {
                this.initMonaco();
            });
        }
        else {
            loadedMonaco = true;
            loadPromise = new Promise((resolve) => {
                let baseUrl = this.config.baseUrl;
                // ensure backward compatibility
                if (baseUrl === "assets" || !baseUrl) {
                    baseUrl = "./assets/monaco/min/vs";
                }
                if (typeof (window.monaco) === 'object') {
                    this.initMonaco();
                    resolve();
                    return;
                }
                const onGotAmdLoader = (require) => {
                    let usedRequire = require || window.require;
                    let requireConfig = { paths: { vs: `${baseUrl}` } };
                    Object.assign(requireConfig, this.config.requireConfig || {});
                    // Load monaco
                    usedRequire.config(requireConfig);
                    usedRequire([`vs/editor/editor.main`], () => {
                        if (typeof this.config.onMonacoLoad === 'function') {
                            this.config.onMonacoLoad();
                        }
                        this.initMonaco();
                        resolve();
                    });
                };
                if (this.config.monacoRequire) {
                    onGotAmdLoader(this.config.monacoRequire);
                    // Load AMD loader if necessary
                }
                else if (!window.require) {
                    const loaderScript = document.createElement('script');
                    loaderScript.type = 'text/javascript';
                    loaderScript.src = `${baseUrl}/loader.js`;
                    loaderScript.addEventListener('load', () => { onGotAmdLoader(); });
                    document.body.appendChild(loaderScript);
                    // Load AMD loader without over-riding node's require
                }
                else if (!window.require.config) {
                    var src = `${baseUrl}/loader.js`;
                    var loaderRequest = new XMLHttpRequest();
                    loaderRequest.addEventListener("load", () => {
                        let scriptElem = document.createElement('script');
                        scriptElem.type = 'text/javascript';
                        scriptElem.text = [
                            // Monaco uses a custom amd loader that over-rides node's require.
                            // Keep a reference to node's require so we can restore it after executing the amd loader file.
                            'var nodeRequire = require;',
                            loaderRequest.responseText.replace('"use strict";', ''),
                            // Save Monaco's amd require and restore Node's require
                            'var monacoAmdRequire = require;',
                            'require = nodeRequire;',
                            'require.nodeRequire = require;'
                        ].join('\n');
                        document.body.appendChild(scriptElem);
                        onGotAmdLoader(window.monacoAmdRequire);
                    });
                    loaderRequest.open("GET", src);
                    loaderRequest.send();
                }
                else {
                    onGotAmdLoader();
                }
            });
        }
    }
    /** Tear down the current editor and create a fresh one. */
    reinit() {
        this.disposeEditor();
        this.initMonaco();
    }
    /** Dispose the editor and every resource it owns (listeners, models, subscriptions). */
    disposeEditor() {
        this._windowResizeSubscription?.unsubscribe();
        this._windowResizeSubscription = undefined;
        this._listeners.forEach(listener => listener.dispose());
        this._listeners = [];
        this._editor?.dispose();
        this._editor = undefined;
        this._models.forEach(model => model.dispose());
        this._models = [];
    }
    ngOnDestroy() {
        this.disposeEditor();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: BaseEditor, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.2.0", version: "22.0.4", type: BaseEditor, isStandalone: false, selector: "ng-component", inputs: { insideNg: { classPropertyName: "insideNg", publicName: "insideNg", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { onInit: "onInit" }, viewQueries: [{ propertyName: "editorContainer", first: true, predicate: ["editorContainer"], descendants: true, isSignal: true }], ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.Eager }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: BaseEditor, decorators: [{
            type: Component,
            args: [{
                    template: '',
                    changeDetection: ChangeDetectionStrategy.Eager,
                    standalone: false
                }]
        }], ctorParameters: () => [], propDecorators: { insideNg: [{ type: i0.Input, args: [{ isSignal: true, alias: "insideNg", required: false }] }], onInit: [{ type: i0.Output, args: ["onInit"] }], editorContainer: [{ type: i0.ViewChild, args: ['editorContainer', { isSignal: true }] }] } });

class EditorComponent extends BaseEditor {
    constructor() {
        super();
        this.zone = inject(NgZone);
        this._value = '';
        this._disabled = false;
        this.options = input({}, /* @ts-ignore */
        ...(ngDevMode ? [{ debugName: "options" }] : /* istanbul ignore next */ []));
        this.model = input(/* @ts-ignore */
        ...(ngDevMode ? [undefined, { debugName: "model" }] : /* istanbul ignore next */ []));
        this.propagateChange = (_) => { };
        this.onTouched = () => { };
        // Re-create the editor whenever the options or model inputs change.
        effect(() => {
            this.options();
            this.model();
            if (this._editor) {
                this.reinit();
            }
        });
    }
    writeValue(value) {
        this._value = value || '';
        // Fix for value change while dispose in process.
        setTimeout(() => {
            if (this._editor && !this.model()) {
                this._editor.setValue(this._value);
            }
        });
    }
    registerOnChange(fn) {
        this.propagateChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    setDisabledState(isDisabled) {
        this._disabled = isDisabled;
        // Apply immediately to a live editor instead of waiting for a re-init.
        this._editor?.updateOptions({ readOnly: isDisabled || !!this.options().readOnly });
    }
    /** Change the global monaco theme (e.g. `'vs-dark'`). */
    setTheme(themeName) {
        monaco.editor.setTheme(themeName);
    }
    initMonaco() {
        const options = {
            ...this.config.defaultOptions,
            ...this.options()
        };
        if (this._disabled) {
            options.readOnly = true;
        }
        const modelInput = this.model();
        if (modelInput) {
            const existing = modelInput.uri ? monaco.editor.getModel(modelInput.uri) : null;
            if (existing) {
                existing.setValue(this._value);
                options.model = existing;
            }
            else {
                const created = monaco.editor.createModel(modelInput.value, modelInput.language, modelInput.uri);
                this._models.push(created);
                options.model = created;
            }
        }
        const container = this.editorContainer().nativeElement;
        if (this.insideNg()) {
            this._editor = monaco.editor.create(container, options);
        }
        else {
            this.zone.runOutsideAngular(() => {
                this._editor = monaco.editor.create(container, options);
            });
        }
        if (!options.model) {
            this._editor.setValue(this._value);
        }
        this._listeners.push(this._editor.onDidChangeModelContent(() => {
            const value = this._editor.getValue();
            // value is not propagated to parent when executing outside zone.
            this.zone.run(() => {
                this.propagateChange(value);
                this._value = value;
            });
        }));
        this._listeners.push(this._editor.onDidBlurEditorWidget(() => {
            this.onTouched();
        }));
        // refresh layout on resize event.
        this._windowResizeSubscription = fromEvent(window, 'resize').subscribe(() => this._editor.layout());
        this.onInit.emit(this._editor);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: EditorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "22.0.4", type: EditorComponent, isStandalone: true, selector: "ngx-monaco-editor", inputs: { options: { classPropertyName: "options", publicName: "options", isSignal: true, isRequired: false, transformFunction: null }, model: { classPropertyName: "model", publicName: "model", isSignal: true, isRequired: false, transformFunction: null } }, providers: [{
                provide: NG_VALUE_ACCESSOR,
                useExisting: forwardRef(() => EditorComponent),
                multi: true
            }], usesInheritance: true, ngImport: i0, template: '<div class="editor-container" #editorContainer></div>', isInline: true, styles: [":host{display:block;height:200px}.editor-container{width:100%;height:98%}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: EditorComponent, decorators: [{
            type: Component,
            args: [{ standalone: true, selector: 'ngx-monaco-editor', template: '<div class="editor-container" #editorContainer></div>', changeDetection: ChangeDetectionStrategy.OnPush, providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(() => EditorComponent),
                            multi: true
                        }], styles: [":host{display:block;height:200px}.editor-container{width:100%;height:98%}\n"] }]
        }], ctorParameters: () => [], propDecorators: { options: [{ type: i0.Input, args: [{ isSignal: true, alias: "options", required: false }] }], model: [{ type: i0.Input, args: [{ isSignal: true, alias: "model", required: false }] }] } });

class DiffEditorComponent extends BaseEditor {
    constructor() {
        super();
        this.zone = inject(NgZone);
        this.options = input({}, /* @ts-ignore */
        ...(ngDevMode ? [{ debugName: "options" }] : /* istanbul ignore next */ []));
        this.originalModel = input(/* @ts-ignore */
        ...(ngDevMode ? [undefined, { debugName: "originalModel" }] : /* istanbul ignore next */ []));
        this.modifiedModel = input(/* @ts-ignore */
        ...(ngDevMode ? [undefined, { debugName: "modifiedModel" }] : /* istanbul ignore next */ []));
        // Re-create the editor whenever options or either model change.
        effect(() => {
            this.options();
            this.originalModel();
            this.modifiedModel();
            if (this._editor) {
                this.reinit();
            }
        });
    }
    initMonaco() {
        const original = this.originalModel();
        const modified = this.modifiedModel();
        if (!original || !modified) {
            throw new Error('originalModel or modifiedModel not found for ngx-monaco-diff-editor');
        }
        const options = {
            ...this.config.defaultOptions,
            ...this.options()
        };
        const originalModel = monaco.editor.createModel(original.code, original.language || options.language);
        const modifiedModel = monaco.editor.createModel(modified.code, modified.language || options.language);
        this._models.push(originalModel, modifiedModel);
        const container = this.editorContainer().nativeElement;
        container.innerHTML = '';
        if (this.insideNg()) {
            this._editor = monaco.editor.createDiffEditor(container, options);
        }
        else {
            this.zone.runOutsideAngular(() => {
                this._editor = monaco.editor.createDiffEditor(container, options);
            });
        }
        this._editor.setModel({
            original: originalModel,
            modified: modifiedModel
        });
        // refresh layout on resize event.
        this._windowResizeSubscription = fromEvent(window, 'resize').subscribe(() => this._editor.layout());
        this.onInit.emit(this._editor);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: DiffEditorComponent, deps: [], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "22.0.4", type: DiffEditorComponent, isStandalone: true, selector: "ngx-monaco-diff-editor", inputs: { options: { classPropertyName: "options", publicName: "options", isSignal: true, isRequired: false, transformFunction: null }, originalModel: { classPropertyName: "originalModel", publicName: "originalModel", isSignal: true, isRequired: false, transformFunction: null }, modifiedModel: { classPropertyName: "modifiedModel", publicName: "modifiedModel", isSignal: true, isRequired: false, transformFunction: null } }, usesInheritance: true, ngImport: i0, template: '<div class="editor-container" #editorContainer></div>', isInline: true, styles: [":host{display:block;height:200px}.editor-container{width:100%;height:98%}\n"], changeDetection: i0.ChangeDetectionStrategy.OnPush }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: DiffEditorComponent, decorators: [{
            type: Component,
            args: [{ standalone: true, selector: 'ngx-monaco-diff-editor', template: '<div class="editor-container" #editorContainer></div>', changeDetection: ChangeDetectionStrategy.OnPush, styles: [":host{display:block;height:200px}.editor-container{width:100%;height:98%}\n"] }]
        }], ctorParameters: () => [], propDecorators: { options: [{ type: i0.Input, args: [{ isSignal: true, alias: "options", required: false }] }], originalModel: [{ type: i0.Input, args: [{ isSignal: true, alias: "originalModel", required: false }] }], modifiedModel: [{ type: i0.Input, args: [{ isSignal: true, alias: "modifiedModel", required: false }] }] } });

class MonacoEditorModule {
    static forRoot(config = {}) {
        return {
            ngModule: MonacoEditorModule,
            providers: [
                { provide: NGX_MONACO_EDITOR_CONFIG, useValue: config }
            ]
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: MonacoEditorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "22.0.4", ngImport: i0, type: MonacoEditorModule, imports: [CommonModule,
            EditorComponent,
            DiffEditorComponent], exports: [EditorComponent,
            DiffEditorComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: MonacoEditorModule, imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: MonacoEditorModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        EditorComponent,
                        DiffEditorComponent
                    ],
                    exports: [
                        EditorComponent,
                        DiffEditorComponent
                    ]
                }]
        }] });
function provideMonacoEditor(config = {}) {
    return makeEnvironmentProviders([
        { provide: NGX_MONACO_EDITOR_CONFIG, useValue: config }
    ]);
}

/**
 * Generated bundle index. Do not edit.
 */

export { DiffEditorComponent, EditorComponent, MonacoEditorModule, NGX_MONACO_EDITOR_CONFIG, provideMonacoEditor };
